pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskGenerator.java

```diff
@@ -0,0 +1,193 @@
+        LOGGER.info("No segments were selected for compaction for table: {}", tableNameWithType);
+        continue;
+      }
+
+      // map each completedSegment to its partition
+      Map<Integer, List<String>> partitionToSegmentNames = new HashMap<>();
+      for (SegmentZKMetadata completedSegment : completedSegments) {
+        String segmentName = completedSegment.getSegmentName();
+        Integer partitionId = SegmentUtils.getRealtimeSegmentPartitionId(
+            segmentName, tableNameWithType, completedSegment);
+
+        Preconditions.checkState(partitionId != null,
+            "Unable to find partitionId for completedSegment");
+
+        partitionToSegmentNames.computeIfAbsent(partitionId, k -> new ArrayList<>()).add(segmentName);
+      }
+      int numTaskConfigsForTable = 0;
+      for (Map.Entry<Integer, List<String>> entry : partitionToSegmentNames.entrySet()) {
+        List<String> completedSegmentNames = entry.getValue();
+        PinotTaskConfig pinotTaskConfig =
+            getPinotTaskConfig(tableNameWithType, compactionConfigs, completedSegmentNames);
+        pinotTaskConfigs.add(pinotTaskConfig);
+        numTaskConfigsForTable++;
+      }
+      LOGGER.info("Finished generating {} tasks configs for table: {} " + "for task: {}",
+          numTaskConfigsForTable, tableNameWithType, taskType);
+    }
+    return pinotTaskConfigs;
+  }
+
+  private static PinotTaskConfig getPinotTaskConfig(String tableNameWithType, Map<String, String> compactionConfigs,
+      List<String> completedSegmentNames) {
+    Map<String, String> configs = new HashMap<>();
+    configs.put(MinionConstants.TABLE_NAME_KEY, tableNameWithType);
+    configs.put(MinionConstants.SEGMENT_NAME_KEY,
+        StringUtils.join(completedSegmentNames, MinionConstants.SEGMENT_NAME_SEPARATOR));
+    // TODO: use default values if the keys below aren't supplied
+    configs.put(UpsertCompactionTask.BUCKET_TIME_PERIOD_KEY,
+        compactionConfigs.get(UpsertCompactionTask.BUCKET_TIME_PERIOD_KEY));
+    configs.put(UpsertCompactionTask.MAX_NUM_RECORDS_PER_SEGMENT_KEY,
+        compactionConfigs.get(UpsertCompactionTask.MAX_NUM_RECORDS_PER_SEGMENT_KEY));
+    configs.put(UpsertCompactionTask.INVALID_RECORDS_THRESHOLD,
+        compactionConfigs.get(UpsertCompactionTask.INVALID_RECORDS_THRESHOLD));
+    PinotTaskConfig pinotTaskConfig = new PinotTaskConfig(UpsertCompactionTask.TASK_TYPE, configs);
+    return pinotTaskConfig;
+  }
+
+  private static final String[] VALID_CONFIG_KEYS = {
+      UpsertCompactionTask.BUCKET_TIME_PERIOD_KEY,
+      UpsertCompactionTask.BUFFER_TIME_PERIOD_KEY,
+      UpsertCompactionTask.MAX_NUM_RECORDS_PER_SEGMENT_KEY,
+      UpsertCompactionTask.INVALID_RECORDS_THRESHOLD,
+  };
+
+  private Map<String, String> getCompactionConfigs(Map<String, String> taskConfig) {
+    Map<String, String> compactionConfigs = new HashMap<>();
+
+    for (Map.Entry<String, String> entry : taskConfig.entrySet()) {
```

Comment:
There's a `subset` method in `PinotConfiguration` which can be helpful here.

Reply:
As `subset` requires a common prefix it doesn't appear to apply in this case.

Reply:
I think the usual approach is to have configs with a common prefix, say `pinot.minion.task.upsert.compaction.A.B.C`, strip out the prefix and then use `A.B.C` directly.

I haven't taken a deep look here but can the `key.endsWith(configKey)` comparison be rewritten to in such a way?

Reply:
It appears that task configs don't follow that naming convention. 

If you take a look into MinionContants.java you'll see that the common task settings are inherited from `MergeTask`.